src/query/functions/aggregation/function.go

```diff
@@ -122,24 +122,39 @@ func stddevFn(values []float64, bucket []int) float64 {
 }
 
 func varianceFn(values []float64, bucket []int) float64 {
-	sum, count := sumAndCount(values, bucket)
-
-	// Cannot take population standard deviation of less than 1 value
-	if count < 1 {
+	if len(values) == 0 || len(bucket) == 0 {
 		return math.NaN()
 	}
 
-	average := sum / count
-	sumOfSquares := 0.0
+	// Using Welford's online algorithm for calculating variance
+	// https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Welford's_online_algorithm
+	//
+	// This algorithm is used in Prometheus and also should provide better numerical precision than
```

Comment:
Code in prometheus:
https://github.com/prometheus/prometheus/blob/e7e60623ff312ba47906fd4af49a7a24eb32f380/promql/engine.go#L2017-L2021
https://github.com/prometheus/prometheus/blob/e7e60623ff312ba47906fd4af49a7a24eb32f380/promql/engine.go#L2062-L2066